hydra-cluster/src/Hydra/Generator.hs

```diff
@@ -80,53 +82,108 @@ defaultProtocolParameters = def
   let fundingTransaction =
-        mkGenesisTx
+        buildRawTransaction
           networkId
+          initialInput
           faucetSk
           (Coin availableInitialFunds)
           clientFunds
-  clientDatasets <- forM clientKeys (generateClientDataset fundingTransaction)
+  let dataset clientKeys =
+        generateClientDataset networkId fundingTransaction clientKeys nTxs generateOneRandomTransfer
+  clientDatasets <- forM allClientKeys dataset
   pure Dataset{fundingTransaction, clientDatasets, title = Nothing, description = Nothing}
  where
-  generateClientDataset fundingTransaction clientKeys@ClientKeys{externalSigningKey} = do
-    let vk = getVerificationKey externalSigningKey
-        keyPair = (vk, externalSigningKey)
-        -- NOTE: The initialUTxO must all UTXO we will later commit. We assume
-        -- that everything owned by the externalSigningKey will get committed
-        -- into the head.
-        initialUTxO =
-          utxoProducedByTx fundingTransaction
-            & UTxO.filter ((== mkVkAddress networkId vk) . txOutAddress)
-    txSequence <-
-      reverse
-        . thrd
-        <$> foldM (generateOneTransfer networkId) (initialUTxO, keyPair, []) [1 .. nTxs]
-    pure ClientDataset{clientKeys, initialUTxO, txSequence}
-
-  thrd (_, _, c) = c
+  initialInput =
+    genesisUTxOPseudoTxIn
+      networkId
+      (unsafeCastHash $ verificationKeyHash $ getVerificationKey faucetSk)
+
+-- | Generate 'Dataset' which does not grow the per-client UTXO set over time.
+-- This queries the network to fetch the current funds available in the faucet
+-- to be distributed among the peers.
+-- The sequence of transactions generated consist only of simple self payments.
+generateDemoUTxODataset ::
+  NetworkId ->
+  SocketPath ->
+  -- | Number of clients
+  [ClientKeys] ->
+  -- | Number of transactions
+  Int ->
+  IO Dataset
+generateDemoUTxODataset network nodeSocket allClientKeys nTxs = do
+  (faucetVk, faucetSk) <- keysFor Faucet
+  faucetUTxO <- queryUTxOFor network nodeSocket QueryTip faucetVk
+  let (Coin fundsAvailable) = selectLovelace (balance @Tx faucetUTxO)
+  -- Prepare funding transaction which will give every client's
+  -- 'externalSigningKey' "some" lovelace. The internal 'signingKey' will get
+  -- funded in the beginning of the benchmark run.
+  clientFunds <- generate $ genClientFunds allClientKeys fundsAvailable
+  fundingTransaction <- do
+    let changeAddress = mkVkAddress network faucetVk
+    let recipientOutputs =
+          flip map clientFunds $ \(vk, ll) ->
+            TxOut
+              (mkVkAddress network vk)
+              (lovelaceToValue ll)
+              TxOutDatumNone
+              ReferenceScriptNone
+    buildTransaction network nodeSocket changeAddress faucetUTxO [] recipientOutputs >>= \case
+      Left e -> throwIO $ FaucetFailedToBuildTx{reason = e}
+      Right body -> do
+        let signedTx = sign faucetSk body
+        pure signedTx
```

Comment:
Could: DRY this with building the funding transaction in `generateConstantUTxODataset`.

We have two variants of dataset returning functions now:
* `generateConstantUTxODataset` uses hard-coded `availableInitialFunds`, `Faucet` keys and generates clients keys
* `generateDemoUTxODataset` that queries available funds takes keys as argument

This could be combined differently such that we could define one in terms of the other. Also, do we still need the hard-coded variant if we would query the devnet there too?

Reply:
good idea.

lets address this in a subsequent refactor.